src/main/java/uk/gov/ch/oauth/Oauth2.java

```diff
@@ -63,51 +58,50 @@ public Oauth2(final IIdentityProvider identityProvider, SessionFactory sessionFa
     public String oauth2EncodeState(final String returnUri,
             final String nonce,
             final String attributeName) {
-
-        final JSONObject payloadJson = new JSONObject();
-        payloadJson.put(attributeName, returnUri);
-        payloadJson.put("nonce", nonce);
-
-        final Payload payload = new Payload(payloadJson);
-        final JWEHeader header = new JWEHeader(JWEAlgorithm.DIR, EncryptionMethod.A128CBC_HS256);
-        final JWEObject jweObject = new JWEObject(header, payload);
-
-        try {
-            final DirectEncrypter encrypter = new DirectEncrypter(identityProvider.getRequestKey());
-            jweObject.encrypt(encrypter);
-        } catch (final JOSEException e) {
-            LOGGER.error("Could not encode OAuth state", e);
-            return null;
-        }
-
-        return jweObject.serialize();
+        return oAuth2StateHandler.oauth2EncodeState(returnUri, nonce, attributeName);
     }
 
-    public String oauth2EncodeState(final String returnUri,
+    @Override
+    public String encodeSignInState(final String returnUri,
             final Session session,
             final String attributeName) {
-        return oauth2EncodeState(returnUri, nonceGenerator.setNonceForSession(session),
-                attributeName);
+        return oAuth2StateHandler
+                .oauth2EncodeState(returnUri, nonceGenerator.setNonceForSession(session),
+                        attributeName);
     }
 
-
     /**
      * Given a state encapsulating a JWE token, decode it into a {@link com.nimbusds.jose.Payload}
      */
     @Override
     public Payload oauth2DecodeState(final String state) {
-        Payload payload;
-        try {
-            final JWEObject jweObject = JWEObject.parse(state);
+        return oAuth2StateHandler.oauth2DecodeState(state);
+    }
 
-            final byte[] key = identityProvider.getRequestKey();
-            jweObject.decrypt(new DirectDecrypter(key));
-            payload = jweObject.getPayload();
-        } catch (final Exception e) {
-            LOGGER.error("Could not decode OAuth state", e);
-            payload = null;
+    @Override
+    public boolean isValid(final String state, final String code) {
+        final String returnedNonce = getNonceFromState(state);
+        boolean valid = oauth2VerifyNonce(returnedNonce);
+        if (!valid) {
+            LOGGER.error("Invalid nonce value in state");
+        } else {
+            valid = extractUserProfile(code);
```

Comment:
I would prefer having a new boolean here as currently the default value of `valid` is true.

Reply:
I understand the reasoning, I've simplified the code to do reduce the conditional nesting.